napari/utils/events/containers/_set.py

```diff
@@ -54,10 +51,16 @@ def __iter__(self) -> Iterator[_T]:
     def __len__(self) -> int:
         return len(self._set)
 
+    def _pre_add_hook(self, value):
+        # for subclasses to potentially check value before adding
+        return value
+
     def add(self, value: _T) -> None:
+        """Add an element to the set, if not already present."""
         if value not in self:
+            value = self._pre_add_hook(value)
             self._set.add(value)
-            self.events.added(value={value})
+            self.events.changed(added={value}, removed={})
```

Comment:
Do we want to emit if it was already there?

Reply:
this doesn't emit if you add something that is already there.  (since "listeners" don't need to take any action).  were you thinking we might want to emit anyway?  or thinking that this _would_ emit anyway?

Reply:
The second one!